vfs/cache.go

```diff
@@ -270,12 +270,63 @@ func (c *cache) exists(name string) bool {
 	if err != nil {
 		return false
 	}
-	if fi.IsDir() {
+	// checks for non-regular files (e.g. directories, symlinks, devices, etc.)
+	if !fi.Mode().IsRegular() {
 		return false
 	}
 	return true
 }
 
+// renames the file in cache
+func (c *cache) rename(name string, newName string) bool {
+	osOldPath := c.toOSPath(name)
+	osNewPath := c.toOSPath(newName)
+	sfi, err := os.Stat(osOldPath)
+	if err != nil {
+		return false
+	}
+	if !sfi.Mode().IsRegular() {
+		// cannot copy non-regular files (e.g., directories, symlinks, devices, etc.)
+		fs.Errorf("rename: non-regular source file %s (%q)", sfi.Name(), sfi.Mode().String())
+		return false
+	}
+	dfi, err := os.Stat(osNewPath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			fs.Errorf(name, "rename: Failed to stat destination: %v", err)
+			return false
+		}
+		parent := findParent(osNewPath)
+		err = os.MkdirAll(parent, 0700)
+		if err != nil {
+			return false
+		}
+	} else {
+		if !(dfi.Mode().IsRegular()) {
+			fs.Errorf("rename: non-regular destination file: %s (%q)", dfi.Name(), dfi.Mode().String())
+			return false
+		}
+		if os.SameFile(sfi, dfi) {
+			return true
+		}
+	}
+	if err = os.Link(osOldPath, osNewPath); err == nil {
```

Comment:
Why not use os.Rename always?  Isn't it equivalent to Link+Remove?

os.Link doesn't work on all windows fileystems (eg FAT)

os.Rename may have troubles renaming over an existing file on Windows.

Reply:
Not much of difference either way. Didn't try it on windows, if it doesn't work there, I will just remove the section that tries to link.

As for rename, what do you think would be better? Remove the existing destination or do nothing and print an error log?